servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultHostAndPort.java

```diff
@@ -62,4 +135,126 @@ public boolean equals(Object o) {
+        if (rawIp.isEmpty()) {
+            throw new IllegalArgumentException("Empty IP");
+        }
+        // https://datatracker.ietf.org/doc/html/rfc5952#section-2
+        // JDK doesn't do IPv6 compression, or remove leading 0s. This may lead to inconsistent String representation
+        // which will yield different hash-codes and equals comparisons to fail when it shouldn't.
+        int longestZerosCount = 0;
+        int longestZerosBegin = -1;
+        int longestZerosEnd = -1;
+        int zerosCount = 0;
+        int zerosBegin = rawIp.charAt(0) != '0' ? -1 : 0;
+        int zerosEnd = -1;
+        int lastColon = -1;
+        boolean isCompressed = false;
+        char prevChar = '\0';
+        StringBuilder compressedIPv6Builder = new StringBuilder(rawIp.length());
+        for (int i = 0; i < rawIp.length(); ++i) {
+            final char c = rawIp.charAt(i);
+            switch (c) {
+                case '0':
+                    if (zerosBegin < 0 || i == rawIp.length() - 1) {
+                        compressedIPv6Builder.append('0');
+                    }
+                    break;
+                case ':':
+                    if (prevChar == ':') {
+                        isCompressed = true;
+                        compressedIPv6Builder.append(':');
+                    } else if (zerosBegin >= 0) {
+                        ++zerosCount;
+                        compressedIPv6Builder.append("0:");
+                        zerosEnd = compressedIPv6Builder.length();
+                    } else {
+                        compressedIPv6Builder.append(':');
+                        zerosBegin = compressedIPv6Builder.length();
+                    }
+                    lastColon = i;
+                    break;
+                default:
+                    if (!isValidateIPv6Digit(c)) {
+                        boolean shouldInsertRemainder = false;
+                        if (c == '.') { // check for IPv4 mapped
+                            final int zoneIdIndex = rawIp.lastIndexOf('%');
+                            validateIPv4(rawIp, lastColon + 1, zoneIdIndex < 0 ? rawIp.length() : zoneIdIndex);
+                            shouldInsertRemainder = true;
+                        } else if (c == '%') { // check for IPv6 zone id
+                            // no constraints enforced on zone id
+                            shouldInsertRemainder = true;
+                        }
+                        if (shouldInsertRemainder) {
+                            compressedIPv6Builder.append(c);
+                            ++i;
+                            for (; i < rawIp.length(); ++i) {
+                                compressedIPv6Builder.append(rawIp.charAt(i));
+                            }
+                            break;
+                        } else {
+                            throw new IllegalArgumentException("Invalid IPv6 address[" + i + "]=" + c);
```

Comment:
Consider including `rawIp`

Reply:
I removed the validation embedded in this method and just used NetworkUtils ... we can combine/optimize later if needed.